### ISSUE TITLE
 Rate Limit Exceeded

### ISSUE THREAD

**Issue by EverMineServer** [prasmussen/gdrive#306]
Is there a way for fix this error?
`Failed to upload file: googleapi: Error 403: Rate Limit Exceeded, rateLimitExceeded
`

**Comment by patbiber** [comment]
You need to request for higher Limits on the Google API.
Over the Google Drive API on the Google Cloud Platform.

**Comment by EverMineServer** [comment]
Can I do it?

**Comment by euklid** [comment]
See #132 
I just opened a pull request because I encountered the same problem and thus I just learnt Go for this purpose ;) @EverMineServer If you only need to use the `download` or `download query` command, then you should be able to use the code in my branch, but no promises until it's actually merged

**Comment by jucor** [comment]
I'm having the same problem, when "gdrive upload --recursive" a directory containing 2000 small files totally a measly 8 MB.

**Comment by euklid** [comment]
@jucor probably you can take a look at my PR #309 and try to apply the same changes for the upload. I'm not using this program any more (only needed it for exactly one occasion for downloading...) and currently don't have time to apply similar changes to the other parts of the program. At least the download did work for me then, when I was downloading over 2000 files.

**Comment by singhravi1** [comment]
Hi,
Any updated solution for this?

**Comment by NickMoutsios** [comment]
I am having the exact same issue when trying to upload a single file. The rate which I am uploading in the past few days is like 20 MB / day, hard to believe I have reached a limit.

**Comment by NickMoutsios** [comment]
Interesting... It just works again, have touched nothing other than visiting the API dashboard to check what is going on. These computers nowadays....

**Comment by singhravi1** [comment]
Yes, this is what is happening, it is giving the error randomly.
This needs to be fixed.

**Comment by javaarchive** [comment]
It happened to me and then I reran the program and everything was fixed.

**Comment by herbat73** [comment]
It blocks me even upload one large file (60mb)

gdrive: 2.1.0
Golang: go1.6
OS/Arch: linux/amd64

**Comment by cnrting** [comment]
+1 rerun is ok

**Comment by andresmitre** [comment]
I don't thinks it's a size limit, I uploaded around 1 GB yesterday, and now I can not upload around 50 MB. PD. I'm in different Internet connection.

gdrive:  2.1.0
OS: linux (MINT)

**Comment by javaarchive** [comment]
For me, this error occured the first time I used gdrive. Rerunning it fixed everything. 
Note: I'm on a G Suite Account

**Comment by andresmitre** [comment]
I've tried in multiple connections but unsuccessful. it sucks...

**Comment by javaarchive** [comment]
Probaly somebody could make a auto restart feature

**Comment by liamellul** [comment]
I encountered the same issue yesterday, used the --recursive option on all files within a folder. the empty folder was put on google drive but none of the files and I received this error. Google API had my rate at 1,000 calls per 100 seconds.

Does anyone know what is making it throw the error?

**Comment by AquariusPower** [comment]
On my script retry loop, I catch the output and `grep` for "Failed.*rateLimitExceeded" and after 10s it tries again until it works, usually doesnt take more than 3 retries :).
It happens with any file sizes apparently.
I think (expect) they may improve the servers to handle more satisfied users soon! xD

**Comment by frankli0324** [comment]
> On my script retry loop, I catch the output and `grep` for "Failed.*rateLimitExceeded" and after 10s it tries again until it works, usually doesnt take more than 3 retries :).
> It happens with any file sizes apparently.
> I think (expect) they may improve the servers to handle more satisfied users soon! xD

if you do
```bash
while 1;do
./gdrive download {asdf} --recursive --skip
grep...|sleep etc
done
```
you are still re-calling the API, so it's still highly possible to get another 302 error

my dirty fix:
edit $GO_DIR/src/.../gdrive/drive/download.go at around line 245
```go
for _, f := range files {
    // Copy args and update changed fields
    newArgs := args
    newArgs.Path = newPath
    newArgs.Id = f.Id
    newArgs.Stdout = false
    retry:
    err = self.downloadRecursive(newArgs)
    if err != nil {
        print("retry after 5 seconds")
        time.Sleep(5*time.Second)
        goto retry
        return err
    }
}
```
dirty, but (kind of) works
still looking for a fix from repo owner
suggestion: use API keys provided from users instead of an app for everyone

**Comment by AquariusPower** [comment]
cool, i will try to run that script thx!
btw, I never use --recursive, I only work with single files recreating each directory remotely as needed :)

**Comment by guylabbe** [comment]
It works when you run the command a second time.

**Comment by arjanna** [comment]
Same problem here, but I don't use recursive and I am not downloading the files: I upload them through
~/gdrive sync upload --keep-largest 

Any help is appreciated!

**Comment by petarov** [comment]
Same problem as @arjanna describes it on my side. However, looking at my sync logs back till 2017, I think this problem started occurring recently on my side. First log entry is from May, 2018 and it seems to come and go. I think this is somehow Google related. I found no explicit info about rate limits for user authorized tokens.

**Comment by arjanna** [comment]
> Same problem as @arjanna describes it on my side. However, looking at my sync logs back till 2017, I think this problem started occurring recently on my side. First log entry is from May, 2018 and it seems to come and go. I think this is somehow Google related. I found no explicit info about rate limits for user authorized tokens.

I am working around the issue and I found a solution here: https://github.com/gdrive-org/gdrive/issues/426 I hope it works for you @petarov 
!
This fixes the problem from my laptop, but I have to upload files from a supercomputer, where I don't have sudo permits, so I can't use this fix.

**Comment by tiomno** [comment]
> Interesting... It just works again, have touched nothing other than visiting the API dashboard to check what is going on. These computers nowadays....

This worked for me!!! 🎉 🎈 🎂 Definitely and odd validation condition on the API. 🤔

**Comment by andresmitre** [comment]
I gave up, I switched to grive2. Is actually the same.

**Comment by AquariusPower** [comment]
@andresmitre I moved from grive2 to gdrive because to upload anything, even 1kb, I had to **always** download about 5MB (the whole remote file list with info to compare locally) and that would spend a lot of my daily quota, and there was absolutely nothing I could do to make it always one way (force upload only only local changes)
